nightfall-optimist/src/services/database.mjs

```diff
@@ -215,9 +223,24 @@ Function to save a (unprocessed) Transaction
 */
 export async function saveTransaction(_transaction) {
   const transaction = { _id: _transaction.transactionHash, ..._transaction, mempool: true };
+  logger.debug(
+    `saving transaction ${transaction.transactionHash}, with layer 1 block number ${_transaction.blockNumber}`,
+  );
+  // there are three possibilities here:
+  // 1) We're just saving a transaction for the first time.  This is fine
+  // 2) We're trying to save a replayed transaction.  This will correctly fail because the _id will be duplicated
+  // 3) We're trying to save a transaction that we've seen before but it was re-mined due to a chain reorg. In
+  //    this case, it's fine, we just update the layer 1 blocknumber and transactionHash to the new values
   const connection = await mongo.connection(MONGO_URL);
   const db = connection.db(OPTIMIST_DB);
-  return db.collection(TRANSACTIONS_COLLECTION).insertOne(transaction);
+  const query = { transactionHash: transaction.transactionHash };
+  const update = { $set: transaction };
+  const existing = await db.collection(TRANSACTIONS_COLLECTION).findOne(query);
+  if (!existing || !existing.blockNumber) {
+    if (existing) transaction.mempool = false; // it's a re-mined transaction.  This will stop block-assembler making a new block
```

Comment:
My understanding is that this line runs IFF the transaction is in the database AND if the `blockNumber === null` (i.e. this is a transaction that has been remined). It's not clear to me why this would guarantee that this means `mempool = false` for this tx. 

It is possible for me to have submitted a transaction onto a chain fork that did not find its way into a block. `mempool = false` should be a flag that is only set as part of `blockProposed` handling or am I missing something.

Reply:
Yes, quite right.  I need to get the block assembler to ignore re-mined transactions.

Reply:
The problem is that we will then see re-mined L2 blocks which originated on the orphaned chain.  These are not necessarily invalid because if there are no L2 blocks on the new chain, then the re-mined blocks could well be re-mined in the correct blockNumberL2 order.  Thus they won't get reverted at layer 1. However, we don't know what order they'll actually turn up in.  Thus we don't know if our re-mined transaction will actually end up being eventually incorporated in a block or not.  So we don't know whether to allow it to be incorporated into a fresh block or not. 
The simplest thing is just to dis-allow re-mined transactions, which is what my code does by pulling them out of the mem-pool.  That may not be optimal though and eventually we may think up a more sophisticated approach

Reply:
Good point.  I've made the mempool logic more subtle by including a check the the transaction submitted handler, which will mark `mempool = false` iff the transaction is a re-mined one that is _also_ already placed in a L2 block.